File: presentation/src/main/java/by/grodno/vasili/presentation/feature/base/ToolbarView.java
package by.grodno.vasili.presentation.feature.base;

import android.app.Activity;
import android.content.Context;
import android.content.res.TypedArray;
import android.databinding.DataBindingUtil;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.constraint.ConstraintLayout;
import android.util.AttributeSet;
import android.view.LayoutInflater;

import by.grodno.vasili.presentation.R;
import by.grodno.vasili.presentation.databinding.ToolbarViewBinding;

/**
 * Top toolbar with customizable title, ok and cancel handlers
 */
public class ToolbarView extends ConstraintLayout {
    private ToolbarViewBinding binding;

    public ToolbarView(Context context) {
        super(context);
        init(context, null);
    }

    public ToolbarView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs);
    }

    private void init(@NonNull Context context, @Nullable AttributeSet attrs) {
        binding = DataBindingUtil.inflate(LayoutInflater.from(getContext()), R.layout.toolbar_view, this, true);
        if (attrs != null) {
            applyAttributes(context, attrs);
        }
        setDefaultListeners((Activity) context);
    }

    private void applyAttributes(@NonNull Context context, @NonNull AttributeSet attrs) {
        TypedArray array = context.obtainStyledAttributes(attrs, R.styleable.ToolbarView);
        try {
            String title = array.getString(R.styleable.ToolbarView_toolbarTitle);
            binding.toolbarTitle.setText(title);
        } finally {
            array.recycle();
        }
    }

    private void setDefaultListeners(Activity context) {
        binding.checkImage.setOnClickListener(v -> context.finish());
        binding.closeImage.setOnClickListener(v -> context.finish());
    }

    /**
     * Set handler for ok icon pressing
     */
    public void setOkListener(OnClickListener okListener) {
        binding.checkImage.setOnClickListener(okListener);
    }

    /**
     * Set handler for cancel icon pressing
     */
    public void setCancelListener(OnClickListener cancelListener) {
        binding.closeImage.setOnClickListener(cancelListener);
    }
}

File: presentation/src/main/java/by/grodno/vasili/presentation/view/ToolbarView.kt
package by.grodno.vasili.presentation.view

import android.app.Activity
import android.content.Context
import android.databinding.DataBindingUtil
import android.support.constraint.ConstraintLayout
import android.util.AttributeSet
import android.view.LayoutInflater
import by.grodno.vasili.presentation.R
import by.grodno.vasili.presentation.databinding.ToolbarViewBinding

/**
 * Top toolbar with customizable title, ok and cancel handlers. </br>
 * Optional attributes for component in XML:
 *
 *  * app:title - title for toolbar (default: empty) </br>
 *  * app:onOkClick - handler for click on check-icon (default: close activity) </br>
 *  * app:onCancelClick - handler for click on close-icon (default: close activity) </br>
 *
 */
class ToolbarView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyle: Int = 0) :
        ConstraintLayout(context, attrs, defStyle) {

    private var binding: ToolbarViewBinding =
            DataBindingUtil.inflate(LayoutInflater.from(getContext()), R.layout.toolbar_view, this, true)

    init {
        setDefaultListeners(context as Activity, binding)
    }

    private fun setDefaultListeners(context: Activity, binding: ToolbarViewBinding) {
        binding.checkImage.setOnClickListener { context.finish() }
        binding.closeImage.setOnClickListener { context.finish() }
    }

    /**
     * Setter for optional attribute app:title - title for toolbar
     */
    @Suppress("unused")
    fun setTitle(title: String) {
        binding.toolbarTitle.text = title
    }

    /**
     * Setter for optional attribute app:onOkClick
     *
     * @param okHandler handler for check-icon pressing
     */
    @Suppress("unused")
    fun setOnOkClick(okHandler: Runnable) {
        binding.checkImage.setOnClickListener { okHandler.run() }
    }

    /**
     * Setter for optional attribute app:onCancelClick
     *
     * @param cancelHandler handler for close-icon pressing
     */
    @Suppress("unused")
    fun setOnCancelClick(cancelHandler: Runnable) {
        binding.closeImage.setOnClickListener { cancelHandler.run() }
    }
}
